### ISSUE TITLE
Make katakana/hiragana + width interchangeable in Japanese (Find)

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/Pale-Moon#379]
Two things should be done here:
- Make Find ignore width of characters (hankaku<->zenkaku), including non-alpha western/eastern characters
- Make katakana/hirigana distinction linked to the "case sensitive" toggle

Relevant thread: https://forum.palemoon.org/viewtopic.php?f=13&t=11327

**Comment by chaoskagami** [comment]
I don't really frequent the forums because I prefer to follow development, but just my two cents on the second part about kana being case-linked: I don't believe that's correct behavior.

As an example of what I mean: Kurasu クラス (katakana) is a class of students, 暮らす/くらす is the verb to live. These are spelled the same in terms of syllables but aren't the same. There's a lot more homonyms than that as well. Katakana is mostly loan-words, wheras Hiragana is mostly native words and is also used for particles.

It *will* make it easier to search when using an IME but you can actually hit many false positives with kana-insensitivity.

**Comment by wolfbeast** [comment]
See, that was as far as I understood how this worked as well.
In the forum thread though it's suggested that "modern" Japanese language used specifically on informal sites mixes up both alphabets very often. Apparently Chrome also does matching with a straight translation of characters, so I gather it would be acceptable to do for in-page searches.
I'll have to bow to the expertise of people actually native in Japanese to give a better answer. (no pun intended)

**Comment by chaoskagami** [comment]
"Modern" as in internet forums like 2ch is like saying words used on reddit that aren't dictionary words should be in Pale Moon's spellcheck dictionary.

Anyways, if Chrome does it, chances are it's okay since Chrome currently leads browser share in Japan.

**Comment by Ordinaryperson2** [comment]
Disclaimer: I'm also not a native speaker, so can't personally speak for whether this is desirable in general, I just posted in that thread to provide some details I knew.

I just took a look at Chrome in my XP VM, and can confirm that it doesn't distinguish between hiragana/katakana. However, it seems to only adjust for half width vs full width one way: If full width is searched, half width will be found. If half width is searched, full width will not be found.

Sounds a bit like an incomplete implementation to me, but perhaps there's a reason? I would speculate that half width is less used, so more likely to be deliberate, but I'm not really sure.

Anyway, Chrome doesn't allow you to search case sensitively, so it's actually not possible to search specifically for one or the other in Chrome, at least without an extension. If people are happy with that then having the choice via checkbox is surely an upgrade.

(As for the dictionary, I'd argue it's not really like that, since a dictionary has to understand individual words rather than a pattern. It's not like we're asking the browser to help spell nonsense words, just find them through substitution.)

That said, I'm not finding a source about what specifically should be interchangeable - the poster of the thread also suggested １２３４５６７８９０＋－＊＝ and more.
From a little testing, ！＠＃＄％＆（） should be too.
I tried looking through both bugs.webkit.org and bugs.chromium.org but couldn't find an initial implementation of this.

**Comment by chaoskagami** [comment]
While we're at it: I took a quick look into things.

The current implementation just converts everything to lowercase then compares text. It uses a functtion in xpcom to do the conversion, ToLowerCase.

What would need to happen is adding another conversion function that handles CJK and rewriting the find bar to use that.

This isn't even getting into encoding. The current ToLowerCase uses C-strings directly passed in via search, so it's likely it doesn't properly account for all of the different possible encodings, unless PM only uses UTF8 internally (does it?)

In a nutshell, it isn't *that* bad. It's not a simple thing, though.

Relevant files:
xpcom/string/src/nsReadableUtils.cpp - String functions
embedding/components/find/src/nsFind.cpp - Find Backend

**Comment by wolfbeast** [comment]
I'm going to close this for now. It's a can of worms and not easy to implement because of the intricacies of Japanese and how this matching is supposed to work. We're still a web browser, after all, not a word processor where this would be more important to get exactly right and intelligent.
